### PR TITLE
Fix exhook ignore message hook

### PR DESCRIPTION
**Pull request by JimMoen**

### CODE REVIEW

File: apps/emqx_exhook/src/emqx_exhook.erl [hunk 1]
```diff
@@ -78,37 +78,50 @@ cast(Hookpoint, Req, [ServerName|More]) ->
...
 call_fold(Hookpoint, Req, AccFun) ->
     FailedAction = emqx_exhook_mngr:get_request_failed_action(),
     ServerNames = emqx_exhook_mngr:running(),
     case ServerNames == [] andalso FailedAction == deny of
         true ->
+            ?LOG(warning, "No available Server. Stop hook chain execute."),
             {stop, deny_action_result(Hookpoint, Req)};
         _ ->
-            call_fold(Hookpoint, Req, FailedAction, AccFun, ServerNames)
+            InitAcc = Req, %% `Req` (includede message.. etc.) as `InitAcc`
```
Comment (HJianBo) [inline apps/emqx_exhook/src/emqx_exhook.erl:91]: Keep the original style?

File: apps/emqx_exhook/src/emqx_exhook_server.erl [hunk 2]
```diff
@@ -266,6 +256,20 @@ inc_metrics(IncFun, Name) when is_function(IncFun) ->
 inc_metrics(Prefix, Name) when is_list(Prefix) ->
     emqx_metrics:inc(list_to_atom(Prefix ++ atom_to_list(Name))).
 
+need_call(Hookpoint, Req, Hooks) ->
+    case maps:get(Hookpoint, Hooks, undefined) of
+        undefined ->
+            false; %% Unhooked
```
Comment (HJianBo) [inline apps/emqx_exhook/src/emqx_exhook_server.erl:262]: nit:
```suggestion
            false; %% Hookpoint is not mounted on this server
```

File: apps/emqx_exhook/src/emqx_exhook.erl [hunk 3]
```diff
@@ -78,37 +78,50 @@ cast(Hookpoint, Req, [ServerName|More]) ->
...
+                    ?LOG_CALL_RESULT(ServerName, "'STOP_AND_RETURN'", "Stop hook chain execution"),
+                    {stop, NewAcc};
+                {ok, NewAccAsNReq} ->
+                    ?LOG_CALL_RESULT(ServerName, "'CONTINUE'", "Continue calling remaining ExHook servers."),
+                    call_fold(Hookpoint, NewAccAsNReq, FailedAction, MergeAccFun, More);
+                ignore ->
+                    ?LOG_CALL_RESULT(ServerName, "'IGNORE'", "Continue calling remaining ExHook servers."),
+                    call_fold(Hookpoint, Acc, FailedAction, MergeAccFun, More)
             end;
+        ignore ->
```
Comment (HJianBo) [inline apps/emqx_exhook/src/emqx_exhook.erl:117]: nit: add some comments here, and bring it to be the first clause?
```suggestion
        ignore -> %% Server is not mounted / or does not care about this hook
```

File: apps/emqx_exhook/src/emqx_exhook.erl [hunk 4]
```diff
@@ -78,37 +78,50 @@ cast(Hookpoint, Req, [ServerName|More]) ->
...
+            ignore;
         _ ->
```
Comment (HJianBo) [inline apps/emqx_exhook/src/emqx_exhook.erl:119]:
```suggestion
        {error, Reason} ->
```

File: apps/emqx_exhook/src/emqx_exhook.erl [hunk 5]
```diff
@@ -78,37 +78,50 @@ cast(Hookpoint, Req, [ServerName|More]) ->
 
 -spec call_fold(atom(), term(), function())
   -> {ok, term()}
-   | {stop, term()}.
+   | {stop, term()}
+   | ignore.
```
Comment (HJianBo) [inline apps/emqx_exhook/src/emqx_exhook.erl:82]: This added return value needs to be handled where the method is called?. i.e:
```erlang
%% emqx_exhook_handler.erl
on_message_publish(...
    case call_fold('message.publish', Req,
                   fun emqx_exhook_handler:merge_responsed_message/2) of
        {StopOrOk, #{message := NMessage}} ->
            {StopOrOk, assign_to_message(NMessage, Message)};
        %% assert the new returned value
        ignore -> {ok, Message}
    end.
```

File: apps/emqx_exhook/src/emqx_exhook.erl (hunk 1 above)
Comment (JimMoen) [reply apps/emqx_exhook/src/emqx_exhook.erl:91]: It's actually an `Accumulator` for `emqx_hook` system.
I'll Keep original style with an extra comment.

**Review by HJianBo — DISMISSED** [review]

**Review by HJianBo — APPROVED** [review]

**Pull request merged**